### ISSUE TITLE
3D Flutter Api Support

### ISSUE THREAD

**Issue by Andy-Python-Programmer** [flutter/flutter#58145]
Hello everyone, it would be nice to have 3D support for flutter. I know that the flutter team is mainly focused on 2d but it would be great to even get a low end 3D api with basic functionality like terrains, camera and an object. 

### Currently
I know that currently we can draw 3D using the flutter canvas which is somewhat not the best way to draw 3D but may be if you release a low end api someone can make a nice 3D engine for it and it will be very useful for the flutter community.

Thank You

**Comment by Andy-Python-Programmer** [comment]
Hello flutter team it would be nice to have c++ open gl view in flutter then we could draw 3D objects. If this feature is available then we already have the low end api then all we need to work on is the 3D Engine or a 3D Game Engine.

**Comment by softmarshmallow** [comment]
as written in under documentation, flutter team is has plans for 3D support.
https://flutter.dev/docs/resources/faq

Today we don’t support for 3D via OpenGL ES or similar. We have long-term plans to expose an optimized 3D API, but right now we’re focused on 2D.

I'm planning on making production level 3D drawing widget for creating next generation applications, such like this example.

If flutter team provides which api you guys will be using it will be helpful to design architecture, so after flutter team supports 3d officially, my work can still be valid and not be trashed.

Thanks!

**Comment by note8g2018** [comment]
> 
> 
> as written in under documentation, flutter team is has plans for 3D support.
> https://flutter.dev/docs/resources/faq
> 
> Today we don’t support for 3D via OpenGL ES or similar. We have long-term plans to expose an optimized 3D API, but right now we’re focused on 2D.
> 
> I'm planning on making production level 3D drawing widget for creating next generation applications, such like this example.
> 
> If flutter team provides which api you guys will be using it will be helpful to design architecture, so after flutter team supports 3d officially, my work can still be valid and not be trashed.
> 
> Thanks!

Could please you tell us more about what you intend to do? may be we can help 😁

**Comment by softmarshmallow** [comment]
Hi, @note8g2018. thanks for the reply. 
[3d-app concept design](https://dribbble.com/shots/13952652-Weather-App-Visual-Concept) - Let's say this is kind of a app we're trying to build (and this is the future of the app). 
This generally has been done by gif, or satatic animations, but as we support device tilt & user dynamics, only with the actual 3D model can accomplish this kind of ui. 


The 3D Widget should have capability of..
- Background transparancy
- Lighting
- API interface for tilting, rotation, etc..
- No physics support (Sure it'll be better to have a physics support by default, but via api, and hand-written animation code can do the magic)
- More lighting options
- Basic callbacks & listeners for the scene.

**Comment by note8g2018** [comment]
> 
> 
> Hi, @note8g2018. thanks for the reply.
> [3d-app concept design](https://dribbble.com/shots/13952652-Weather-App-Visual-Concept) - Let's say this is kind of a app we're trying to build (and this is the future of the app).
> This generally has been done by gif, or satatic animations, but as we support device tilt & user dynamics, only with the actual 3D model can accomplish this kind of ui.
> 
> The 3D Widget should have capability of..
> 
>     * Background transparancy
> 
>     * Lighting
> 
>     * API interface for tilting, rotation, etc..
> 
>     * No physics support (Sure it'll be better to have a physics support by default, but via api, and hand-written animation code can do the magic)
> 
>     * More lighting options
> 
>     * Basic callbacks & listeners for the scene.

Wow very nice design 😍, how about you join us in this discord server
@everyone  Welcome to our new discord server and bring your friends 
https://discord.gg/TWnnBmS

**Comment by softmarshmallow** [comment]
> > Hi, @note8g2018. thanks for the reply.
> > [3d-app concept design](https://dribbble.com/shots/13952652-Weather-App-Visual-Concept) - Let's say this is kind of a app we're trying to build (and this is the future of the app).
> > This generally has been done by gif, or satatic animations, but as we support device tilt & user dynamics, only with the actual 3D model can accomplish this kind of ui.
> > The 3D Widget should have capability of..
> > ```
> > * Background transparancy
> > 
> > * Lighting
> > 
> > * API interface for tilting, rotation, etc..
> > 
> > * No physics support (Sure it'll be better to have a physics support by default, but via api, and hand-written animation code can do the magic)
> > 
> > * More lighting options
> > 
> > * Basic callbacks & listeners for the scene.
> > ```
> 
> Wow very nice design 😍, how about you join us in this discord server
> @everyone Welcome to our new discord server and bring your friends
> https://discord.gg/TWnnBmS

Well I've joined it, is it a official flutter support channel? we'll need more details for that link

**Comment by Andy-Python-Programmer** [comment]
@softmarshmallow join the official Flutter’s discord server! https://discord.gg/rflutterdev

**Comment by srix55** [comment]
Any updates on this? I guess there's nothing planned as of now and shouldn't expect anything for a few years.

**Comment by shujaatak** [comment]
I am thinking to replace QML with Flutter once Flutter starts supported 3D API.

**Comment by perbone** [comment]
Any updates on this? Does the new rendering paves the way for 3D support? Thanks!

**Comment by note8g2018** [comment]
> Any updates on this? Does the new renderer paves the way for 3D support? Thanks!

I do not think that Flutter will support 3D like OpenGL but may be something like 2.5D

**Comment by curiouscod3** [comment]
Flutter is 2d only. I don't expect 3d engine. That is why I QUIT flutter.

**Comment by yangjiejie12308** [comment]
2023！ https://medium.com/flutter/whats-next-for-flutter-b94ce089f49c

**Comment by kaljitism** [comment]
Is there any update??

**Comment by EiTaNBaRiBoA** [comment]
I've been experimenting with 3D in Flutter, and I've previously used Unity3D for creating VR/AR Apps and Games. One of the major challenges with 3D or world design is accurately placing objects without having to wait for a hot reload or enlarging the codebase, particularly when it comes to colliders. VR/AR development becomes incredibly difficult when managing the XYZ coordinates and trying to visualize the position of all objects without altering the camera perspective in the code. This proves to be an extremely challenging task.

Proposal

Hello, I'd like to propose a feature: the addition of a scene view similar to Unity/Godot. This would allow us to drag and move objects to different locations, enabling a more flexible and efficient design process for world-building and app development. It could significantly streamline the development of 3D games and XR apps.

Key Benefits of an Editor Scene View:

It simplifies the creation of VR/AR apps in a 3D environment.

When working with 2D games, there's no need for an additional third-party app to set up colliders or view your entire map. This is crucial for understanding your world spatially and for effective level design.

Other Advantages of an Editor Scene View:

Navigation: Scene view allows for easy navigation of your scene in 3D or 2D space. You can pan, zoom, and orbit around your scene from any perspective.

Selection & Manipulation: You can select game objects directly in the scene view and adjust their position, rotation, and scale using the transform tools.

Debugging: Scene view enables real-time inspection and manipulation of your game, which is particularly beneficial for debugging.

Component Visualization: Many components and systems provide visual feedback in the scene view, aiding in adjustments.

Prefab Editing: With the prefab mode, you can edit your prefabs directly in the scene view.

Gizmos and Icons: Scene view provides gizmos and icons to represent non-visible objects, simplifying their management.

Terrain Editing: Scene view includes a suite of terrain editing tools for 3D games, allowing direct terrain manipulation.

Summary
An editor scene view could increase development speed and boost Flutter's popularity within the game dev and VR/AR communities.

**Comment by jonahwilliams** [comment]
This was added: https://medium.com/flutter/whats-new-in-flutter-3-24-6c040f87d1e4
